ignite/metrics/metric.py

```diff
@@ -0,0 +1,66 @@
+from abc import ABCMeta, abstractmethod
+
+from ignite.engine import Events
+
+
+class Metric(object):
+    __metaclass__ = ABCMeta
+
+    """
+    Base class for all Metrics.
+
+    Metrics provide a way to compute various quantities of interest in an online
+    fashion without having to store the entire output history of a model.
+    """
+    def __init__(self):
+        self.reset()
+
+    @abstractmethod
+    def reset(self):
+        """
+        Resets the metric to to it's initial state.
+
+        This is called at the start of each evaluation run.
+        """
+        pass
+
+    @abstractmethod
+    def update(self, output):
+        """
+        Updates the metric's state using the passed batch output.
+
+        This is called once for each batch of each evaluation run.
+
+        Args:
+            output: the is the output from the evaluator's process function
+        """
+        pass
+
+    @abstractmethod
+    def compute(self):
+        """
+        Computes the metric based on it's accumulated state.
+
+        This is called at the end of each evaluation run.
+
+        Returns:
+            Any: the actual quantity of interest
+
+        Raises:
+            NotComputableError: raised when the metric cannot be computed
+        """
+        pass
+
+    def started(self, engine, state):
+        self.reset()
+
+    def iteration_completed(self, engine, state):
+        self.update(state.output)
+
+    def completed(self, engine, state):
+        state.metrics.append(self.compute())
+
+    def attach(self, engine):
```

Comment:
what about having an interface similar to the timer here ([see here](https://github.com/pytorch/ignite/blob/master/ignite/handlers/timing.py#L84)) i.e.:

```python
def attach(start=Events.STARTED, update=Events.ITERATION_COMPLETED, compute=Events.COMPLETED, reset=None)
```

Reply:
this way we wouldn't need `started()` and `reset()` or `iteration_completed()` and `update()`,

Reply:
I'm not sure if the flexibility of passing in the events is something we need (or even want) for metrics. Do you have a use case in mind? My goal here was to have a well-defined interface for metrics (`reset`, `update`, `compute`) that doesn't require that developers of new metrics concern themselves with events at all. Getting rid of `started` or `iteration_completed` would mean that event concepts would leak into the metric implementations. What do you think?

Reply:
I agree the defined interface of `reset`, `update` and `compute` is really nice and clean. 

Is it the idea that metrics will always be updated every iteration, and only computed at the end of your engine run as I imagine that metrics will also be used in training (not just evaluation) to monitor. Also, in this case when/how should `reset()` be called? 

I guess the notion of "attaching" the metric to an engine already ties it to an event model, but by adding in the flexibility it allows them to be used in training (i.e. reset every epoch/compute every epoch) and the evaluator. By defining the defaults to something sensible means for the common use case of an evaluator it's just an out-of-the-box `metric.attach(evaluator)`?

It's not super clear with the current API how to use it with a Trainer, or if this is a use case?

Reply:
Ahh I see. The issue here is that I have built this out (so far at least) to be purely used with evaluators. So i guess we should first take a step back and discuss that...

Should these be just used with evaluators or should they work with trainers as well?

My personal preference is for metrics to be solely an evaluation concern. I think using them with trainers will cause quite a bit of confusion. We would need to modify `create_supervised_trainer` to take a set of metrics and change the return value of the supervised `process_fn`. I think the division of Trainer and Evaluator starts to get a lil confusing if metrics are used on both.

The general (supervised) workflow I've had in mind, is you can log the loss returned by the trainer's `process_fn` to gauge convergence every iteration. But evaluation of metrics happens against a static target (no changing weights during the computation) every so often. The evaluation can certainly be done against the training set as well as an validation set as in the example of above. 

What do you think?

Reply:
Ah I see what you mean, yes! I really like this approach! If the user wants to run metrics on the training set they can define an evaluator and run it on the training set. 

Perfect, this looks pretty good to me. Thanks @jasonkriss!